src/misc/func.rs

```diff
@@ -173,8 +173,38 @@ fn catflip(cws: &[f64], r: f64) -> Option<usize> {
     }
 }
 
+pub fn pflip(weights: &[f64], rng: &mut impl Rng) -> usize {
```

Comment:
Can we implement this with a `sum: Option<f64>` argument? We often know the sum of the weights, so we often only need one scan to compute the sum of a predicate.

```rust
pub fn pflip(weights: &[f64], sum: Option<f64>, rng: &mut impl Rng) -> usize {
    let u = sum.unwrap_or_else(|| weights.iter().sum::<f64>()) * rng.gen::<f64>();
    weights.iter().map(|w| if w < u { 1 } else { 0 }).sum::<usize>() // is this count right?
}
```